lib/private/Files/Cache/FileAccess.php

```diff
@@ -94,4 +95,145 @@ public function getByFileIdsInStorage(array $fileIds, int $storageId): array {
 		$rows = $query->executeQuery()->fetchAll();
 		return $this->rowsToEntries($rows);
 	}
+
+	/**
+	 * Retrieves files stored in a specific storage that have a specified ancestor in the file hierarchy.
+	 * Allows filtering by mime types, encryption status, and limits the number of results.
+	 *
+	 * @param int $storageId The ID of the storage to search within.
+	 * @param int $folderId The file ID of the ancestor to base the search on.
+	 * @param int $fileIdCursor The last processed file ID. Only files with a higher ID will be included. Defaults to 0.
+	 * @param int $maxResults The maximum number of results to retrieve. If set to 0, all matching files will be retrieved.
+	 * @param list<int> $mimeTypeIds An array of mime types to filter the results. If empty, no mime type filtering will be applied.
+	 * @param bool $endToEndEncrypted Whether to include EndToEndEncrypted files
+	 * @param bool $serverSideEncrypted Whether to include ServerSideEncrypted files
+	 * @return \Generator A generator yielding matching files as cache entries.
+	 * @throws \OCP\DB\Exception
+	 */
+	public function getByAncestorInStorage(int $storageId, int $folderId, int $fileIdCursor = 0, int $maxResults = 100, array $mimeTypeIds = [], bool $endToEndEncrypted = true, bool $serverSideEncrypted = true): \Generator {
+		$qb = $this->getQuery();
+		$qb->selectFileCache();
+		$qb->andWhere($qb->expr()->eq('filecache.fileid', $qb->createNamedParameter($folderId, IQueryBuilder::PARAM_INT)));
+		$result = $qb->executeQuery();
+		/** @var array{path:string}|false $root */
+		$root = $result->fetch();
+		$result->closeCursor();
+
+		if ($root === false) {
+			throw new Exception('Could not fetch storage root');
+		}
+
+		$qb = $this->getQuery();
+
+		$path = $root['path'] === '' ? '' : $root['path'] . '/';
+
+		$qb->selectDistinct('*')
+			->from('filecache', 'f')
+			->andWhere($qb->expr()->like('f.path', $qb->createNamedParameter($path . '%')))
+			->andWhere($qb->expr()->eq('f.storage', $qb->createNamedParameter($storageId)))
+			->andWhere($qb->expr()->gt('f.fileid', $qb->createNamedParameter($fileIdCursor, IQueryBuilder::PARAM_INT)));
+
+		if (!$endToEndEncrypted) {
+			// End to end encrypted files are descendants of a folder with encrypted=1
+			// Use a subquery to check the `encrypted` status of the parent folder
+			$subQuery = $this->getQuery()->select('p.encrypted')
+				->from('filecache', 'p')
+				->andWhere($qb->expr()->eq('p.fileid', 'f.parent'))
+				->getSQL();
+
+			$qb->andWhere(
+				$qb->expr()->eq($qb->createFunction(sprintf('(%s)', $subQuery)), $qb->createNamedParameter(0, IQueryBuilder::PARAM_INT))
```

Comment:
```suggestion
				$qb->expr()->eq($qb->createFunction($subQuery), $qb->createNamedParameter(0, IQueryBuilder::PARAM_INT))
```
Is this actually necessary? I thought the query builder would wrap the function in parenthesis already.